flink/core/src/main/java/cz/o2/proxima/flink/core/batch/OffsetTracking.java

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2017-2021 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.flink.core.batch;
+
+import cz.o2.proxima.direct.batch.BatchLogObserver;
+import cz.o2.proxima.direct.batch.BatchLogReader;
+import cz.o2.proxima.direct.batch.ObserveHandle;
+import cz.o2.proxima.direct.batch.Offset;
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.storage.StreamElement;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class OffsetTracking {
```

Comment:
Is this something that should be moved to `proxima-direct-core`?

Reply:
Looks completely unrelated to flink.

Reply:
I think we maybe able to integrate this directly into BatchLogObserver interface, without touching the implementations.

Reply:
For now, due to recent discovery of race conditions when committing, I'd rather leave this internal for flink data operator for now. We can move this anytime, once we have a use case / are confident enough with interface stability.

Reply:
Does that issue affect other operators? If so, can you please file a github issue and summarize what you have found?